src/twisted/protocols/_smb/_base.py

```diff
@@ -0,0 +1,396 @@
+
+
+def _conv_arg(obj: Any, attrib: attr.Attribute) -> Any:
+    val = getattr(obj, attrib.name)
+    if type(val) is uuid_mod.UUID:
+        val = val.bytes_le
+    if isinstance(val, enum.Enum):
+        val = val.value
+    return val
+
+
+IGNORE = 0
+ERROR = 1
+OFFSET = 2
+DATA = 3
+
+
+def unpack(cls: Any, data: bytes, offset: int = 0, remainder: int = IGNORE) -> Any:
+    """
+    unpack binary data into an object.
+
+    @param cls: the class, must be decorated with L{attr.s} and have
+    members with the appropriate metadata using the helpers from this
+    module.
+    @type cls: L{type}
+
+    @param data: the data to unpack
+    @type data: L{bytes}
+
+    @param remainder: what to do with remaining data if longer than required
+    to fill C{cls}
+     - C{IGNORE} ignore it
+     - C{ERROR} throw a L{SMBError}
+     - C{OFFSET} return offset into data
+     where remainder begins as second item of tuple
+     - C{DATA} return remaining data as second item of tuple
+    @type remainder: L{int}
+
+    @param offset: offset into data to begin from
+    @type offset: L{int}
+
+    @return: an instance of C{cls}, or a 2-tuple, first item the former,
+             second as determined by C{remainder}
+    """
+    strct = _get_struct(cls)
+    if remainder == ERROR and strct.size + offset < len(data):
+        raise SMBError("unexpected remaining data")
+    ret = strct.unpack_from(data, offset=offset)
+    fields = smb_fields(cls)
+    assert len(fields) == len(ret)
+    kwargs = {}
+    for i in range(len(ret)):
+        val = ret[i]
+        if fields[i].type is uuid_mod.UUID:
+            val = uuid_mod.UUID(bytes_le=val)
+        kwargs[fields[i].name] = val
+    obj = cls(**kwargs)
+    if remainder <= ERROR:
+        return obj
+    elif remainder == OFFSET:
+        return (obj, offset + strct.size)
+    else:
+        return (obj, data[offset + strct.size :])
+
+
+def smb_fields(cls: Any) -> List[Any]:
+    return [i for i in attr.fields(cls) if SMB_METADATA in i.metadata]
+
+
+def _get_struct(cls: Any) -> struct.Struct:
+    try:
+        # we use classes to hold cache of Structs as precompiling is more
+        # efficient
+        strct = cast(struct.Struct, cls._struct)
+    except AttributeError:
+        strct = struct.Struct(
+            "<" + "".join(i.metadata[SMB_METADATA] for i in smb_fields(cls))
+        )
+        cls._struct = strct
+    return strct
+
+
+def calcsize(cls: Any) -> int:
+    """
+    return the size of a structure.
+    @param cls: C{attr} decorated class
+    @rtype: L{int}
+    """
+    strct = _get_struct(cls)
+    return strct.size
+
+
+_leint = struct.Struct("<I")
+
+
+def int32key(d: Dict[int, Any], val: Any) -> int:
+    """
+    generate a new random key for a dictionary
+    @param d: dictionary with 32-bit int keys
+    @type d: L{dict}
+    @param val: new dictionary value
+    @rtype: L{int}
+    """
+    assert len(d) < 0xC0000000  # otherwiae dict so big hard to find keys
```

Comment:
In the rare case that this assertion is actually tripped, let's give the poor sysadmin an error message to look at ;)

```suggestion
    assert len(d) < 0xC0000000, "dictionary too large to find non-conflicting keys"
```